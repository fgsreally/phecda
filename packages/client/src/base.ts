import type { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios'
import type { P, ToClientInstance } from 'phecda-server'
import type { RequestArgs } from './helper'
import { toReq } from './helper'

export function createBeacon(baseUrl: string) {
  return (arg: any) => {
    const { url, body } = toReq(arg as any)

    navigator.sendBeacon(`${baseUrl}${url}`, JSON.stringify(body))
  }
}

export function createReq(instance: AxiosInstance): <R>(arg: R, config?: AxiosRequestConfig) => Promise<AxiosResponse<Awaited<R>>> {
  return (arg: any, config?: AxiosRequestConfig) => {
    const { url, body, method, headers, query } = toReq(arg as RequestArgs)
    if (!method)
      throw new Error ('methods without route decorator won\'t send request')

    const ret = [url] as any[]

    if (Object.keys(body).length > 0)
      ret.push(body)

    ret.push(addToConfig(config, { headers, params: query }))
    // @ts-expect-error misdirction
    return instance[method](...ret)
  }
}

export function createParallelReq(instance: AxiosInstance, route = '/__PHECDA_SERVER__'): < R extends unknown[]>(args: R, config?: AxiosRequestConfig) => Promise<AxiosResponse<{
  [K in keyof R]: Awaited<R[K]> | P.Error
}>> {
  // @ts-expect-error misdirction
  return (args: RequestArgs[], config?: AxiosRequestConfig) => {
    return instance.post(route, args, config)
  }
}

export function isError<T = any>(data: T | P.Error): data is P.Error {
  return typeof data === 'object' && (data as any).__PS_ERROR__
}

export function useC<T extends new (...args: any) => any>(Module: T): ToClientInstance<InstanceType<T>> {
  return new Module()
}

function addToConfig(origin: any, config: Record<string, any>) {
  if (origin) {
    for (const key in config)
      origin[key] = config[key]
  }
  else {
    origin = config
  }
  return origin
}

// // work for vue-request
// // phecda-client request instance + method from class => async function
// export function toAsync<F extends (...args: any) => any>(pcRequest: ReturnType<typeof createReq>, cb: F): (...params: Parameters<F>) => ReturnType<F> {
//   // @ts-expect-error misdirct
//   return async (...params: Parameters<F>) => {
//     return (await pcRequest(cb(...params as any))).data as any
//   }
// }

// type MergedReqArg = Pick<RequestArgs, 'body' | 'query' | 'params' | 'tag' | 'headers'>

//  const merge = (...args: RequestArgs[]) => {
//   const ret = [] as MergedReqArg[]
//   for (const i of args) {
//     const { body, query, params, tag, headers } = i
//     ret.push({ tag, body, query, params, headers })
//   }

//   return ret
// }

//  type RequestMethod = <F extends (...args: any[]) => any >(fn: F, args: Parameters<F>) => Promise<ReturnType<F>>
