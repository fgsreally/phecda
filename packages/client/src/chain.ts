import type { AxiosInstance, AxiosRequestConfig } from 'axios'
import type { ToClientMap } from 'phecda-server'
import { createParallelReq, createReq, isError, useC } from './base'

 type ChainRequester<T extends Record<string, any>> = ToClientMap<T> & {
   options(config: AxiosRequestConfig): ChainRequester<T>
 }

let batchStack: any[] | null
let batchPromise: any

export function createChainReq<C extends Record<string, any>>(instance: AxiosInstance, controllers: C, options?: { batch?: boolean;route?: string }): ChainRequester<C> {
  const rc: any = {
    options(config: AxiosRequestConfig) {
      this._options = config
      return this
    },
  }
  let $r: ReturnType<typeof createReq>
  let $pr: ReturnType<typeof createParallelReq>
  if (options?.batch)
    $pr = createParallelReq(instance, options.route)
  else
    $r = createReq(instance)

  for (const key in controllers) {
    const proxy = new Proxy(useC(controllers[key]), {
      get(target: any, p) {
        const generator = target[p]
        if (typeof generator !== 'function')
          throw new Error(`'${p as string}' on controller must be a function !`)
        return async (...args: any) => {
          const params = generator(...args)

          if (!options?.batch) {
            const { data } = await $r(params, rc._options)
            return data
          }
          else {
            let index: number
            if (!batchStack) {
              index = 0
              batchStack = [params]
              // eslint-disable-next-line no-async-promise-executor
              batchPromise = new Promise(async (resolve) => {
                await Promise.resolve()
                const { data } = await $pr(batchStack!)
                batchStack = []
                resolve(data)
              })
            }
            else {
              index = batchStack.push(params) - 1
            }
            return new Promise((resolve, reject) => {
              batchPromise.then((data: any[]) => {
                batchStack = null
                const ret = data[index]

                if (isError(ret))
                  reject(ret)

                else
                  resolve(ret)
              })
            })
          }
        }
      },
    })
    rc[key] = proxy
  }
  return rc
}
