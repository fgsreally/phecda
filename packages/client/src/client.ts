import type { Construct, ToClientMap } from 'phecda-server'

import { isError } from './base'

// type ChainRequester<T extends Record<string, any>> = ToClientMap<T> & {
//     options(config: AxiosRequestConfig): ChainRequester<T>
// }
export type RequestArg = {
  method: string
  url: string
  query: Record<string, string>
  body?: any
  headers: Record<string, string>
  params: Record<string, string>
} & Record<string, any>

const rawFetch = async ({ url, method, body, headers, query }: RequestArg) => {
  const queryStr = Object.entries(query).map(([k, v]) => `${k}=${v}`).join('&')

  const res = await fetch(`${url}${queryStr ? `?${queryStr}` : ''}`, {
    method,
    body,
    headers: new Headers(headers),

  })

  if (res.headers.get('Content-Type') === 'application/json')
    return res.json()

  return res.text()
}

export function createClient<Controllers extends Record<string, Construct>>(controllers: Controllers, options: {
  fetch?: (arg: RequestArg) => Promise<any>
  batch?: boolean
} = {}): ToClientMap<Controllers> {
  const client: any = {

  }
  let batchStack: any[] | null
  let batchPromise: any

  const { batch, fetch = rawFetch } = options

  for (const key in controllers) {
    const proxy = new Proxy(new controllers[key](), {
      get(target: any, p) {
        const generator = target[p]
        if (typeof generator !== 'function')
          throw new Error(`'${p as string}' on controller must be a function !`)
        return async (...args: any) => {
          const requestArg = generator(...args)

          for (const i in requestArg.params)
            requestArg.url = requestArg.url.replace(`{{${i}}}`, requestArg.params[i])

          if (!batch) {
            return fetch(requestArg)
          }
          else {
            let index: number
            if (!batchStack) {
              index = 0
              batchStack = [requestArg]
              // eslint-disable-next-line no-async-promise-executor
              batchPromise = new Promise(async (resolve) => {
                await Promise.resolve()
                const { data } = await fetch({
                  body: batchStack,
                  url: '/',
                  method: 'post',
                  query: {},
                  params: {},
                  headers: {},
                })
                batchStack = []
                resolve(data)
              })
            }
            else {
              index = batchStack.push(requestArg) - 1
            }
            return new Promise((resolve, reject) => {
              batchPromise.then((data: any[]) => {
                batchStack = null
                const ret = data[index]

                if (isError(ret))
                  reject(ret)

                else
                  resolve(ret)
              })
            })
          }
        }
      },
    })
    client[key] = proxy
  }
  return client
}
