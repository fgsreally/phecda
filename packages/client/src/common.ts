export const SERIES_SYMBOL = '__symbol_series__'
