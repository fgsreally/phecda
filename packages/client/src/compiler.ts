import type { P } from 'phecda-server'

export class Compiler {
  classMap: Record<string, { [key: string]: string }> = {}
  constructor() { }

  getContent() {
    let content = ''

    for (const name in this.classMap) {
      content += `
        export class ${name}{
            ${Object.values(this.classMap[name]).reduce((p, c) => p + c)}
            }`
    }
    return content
  }

  createRequest() {
    let content = 'import {useC} from \'phecda-server\'\n'
    for (const name in this.classMap)
      content += `export const {${Object.keys(this.classMap[name]).join(',')}}=useC(${name})\n`
    return content
  }

  addMethod(args: P.Meta) {
    const {
      route: {
        route = '/',
        type = 'get',
      } = {}, name, method, params, tag,
    } = args
    const url = route.replace(/\/\:([^\/]*)/g, '')
    if (!this.classMap[name])
      this.classMap[name] = {}
    this.classMap[name][method] = `
    ${method}(${genParams(params)}){
const ret={tag:"${tag}-${method}",body:{},headers:{},query:{},params:{},realParam:'',method:"${type}",url:"${url}"}
${params.reduce((p, c, i) => `${p}if(arg${i}!==undefined&&arg${i}!==null){ret.${c.type}${c.key ? `['${c.key}']` : ''}=arg${i}\n${c.type === 'params' ? `ret.realParam+='/'+arg${i}` : ''}}\n`, '')}
return ret
    }
    `
  }
}

function genParams(decorators: any[]) {
  return decorators.map((_, i) => {
    return `${`arg${i}`}`
  }).join(',')
}
