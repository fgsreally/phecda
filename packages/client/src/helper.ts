import type { BaseRequestType } from 'phecda-server'

export function toReq(arg: RequestArgs) {
  const { body, query, method, url, headers } = arg

  return { headers, method, url, body, query }
}

export interface RequestArgs {
  body: Record<string, any>
  headers: Record<string, string>
  query: Record<string, string>
  params: Record<string, string>
  method: BaseRequestType
  url: string
  tag: string
  func: string
  args: any[]
}
