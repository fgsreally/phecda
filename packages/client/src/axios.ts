import type { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios'
import type { P, RequestType } from 'phecda-server'
import { SERIES_SYMBOL } from './common'

interface RequestArgs {
  body: Record<string, any>
  query: Record<string, string>
  params: Record<string, string>
  realParam: string
  method: RequestType
  url: string
  tag: string
}
type MergedReqArg = Pick<RequestArgs, 'body' | 'query' | 'params' | 'tag' >
export function toReq(arg: RequestArgs) {
  const { body, query, realParam, method, url } = arg
  return { method, url, body, query: Object.keys(query).length > 0 ? `?${Object.entries(query).map(([k, v]) => `${k}=${v}`).join('&')}` : '', params: realParam }
}

export const merge = (...args: RequestArgs[]) => {
  const ret = [] as MergedReqArg[]
  for (const i of args) {
    const { body, query, params, tag } = i
    ret.push({ tag, body, query, params })
  }

  return ret
}

export type RequestMethod = <F extends (...args: any[]) => any >(fn: F, args: Parameters<F>) => Promise<ReturnType<F>>

export function createReq(instance: AxiosInstance): <R>(arg: R, config?: AxiosRequestConfig) => Promise<AxiosResponse<P.Res<Awaited<R>>> > {
  // @ts-expect-error methods without route decorator won't send request
  return (arg: any, config?: AxiosRequestConfig) => {
    const { url, params, query, body, method } = toReq(arg as RequestArgs)
    if (!method) {
      console.warn('methods without route decorator won\'t send request')
      return
    }

    const ret = [`${url}${params}${query}`] as any[]
    body && ret.push(body)
    config && ret.push(config)
    // @ts-expect-error misdirction
    return instance[method](...ret)
  }
}

export function createSeriesReq(instance: AxiosInstance, key = '/__PHECDA_SERVER__'): < R extends unknown[]>(args: R, config?: AxiosRequestConfig) => Promise<AxiosResponse<P.ResOrErr<P.Res<R>>>> {
  // @ts-expect-error misdirction
  return (args: RequestArgs[], config?: AxiosRequestConfig) => {
    return instance.post(key, {
      category: 'series',
      data: merge(...args),
    }, config)
  }
}

export function createParallelReq(instance: AxiosInstance, key = '/__PHECDA_SERVER__'): < R extends unknown[]>(args: R, config?: AxiosRequestConfig) => Promise<AxiosResponse<P.ResOrErr<P.Res<R>>>> {
  // @ts-expect-error misdirction
  return (args: RequestArgs[], config?: AxiosRequestConfig) => {
    return instance.post(key, {
      category: 'parallel',
      data: merge(...args),
    }, config)
  }
}

export function isError<T = any>(data: T | P.Error): data is P.Error {
  return typeof data === 'object' && (data as any).error
}

export function $S(index: number, key = ''): any {
  return `${SERIES_SYMBOL}@${index}@${key}`
}

// work for vue-request
// phecda-client request instance + method from class => async function
export function toAsync<F extends (...args: any) => any>(pcRequest: ReturnType<typeof createReq>, cb: F) {
  return async (...params: Parameters<F>) => {
    return (await pcRequest(cb(...params as any) as ReturnType<F>)).data
  }
}
