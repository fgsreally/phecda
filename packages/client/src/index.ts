import { toReq } from './axios'

export function createBeacon(baseUrl: string) {
  return (arg: any) => {
    const { url, body } = toReq(arg as any)

    navigator.sendBeacon(`${baseUrl}${url}`, JSON.stringify(body))
  }
}

export type ExcludeNever<T> = { [K in keyof T as T[K] extends never ? never : K]: T[K] }

export type PickJsonData<T> = ExcludeNever< T extends object | any[]
  ? {
      [K in keyof T]: T[K] extends ((...args: any[]) => any) | Symbol | undefined ? never : PickJsonData<T[K]>;
    }
  : T>

export type GetFnFromObj<O> = Pick<O, {
  [Key in keyof O]: O[Key] extends (...args: any) => any ? Key : never
}[keyof O]>

// exclude function/symbol/undefined in response obj,to make better type intelligence
export type AsyncReturnToJson<T> = ExcludeNever<{
  [Key in keyof T]: T[Key] extends (...args: any) => any ? (...args: Parameters<T[Key]>) => Promise<PickJsonData<Awaited< ReturnType<T[Key]>>>> : never
}>

export type ExcludeNotFn<T> = ExcludeNever<{
  [Key in keyof T]: T[Key] extends (...args: any) => any ? T[Key] : never
}>

export function useC<T extends new (...args: any) => any>(Module: T): ExcludeNotFn<InstanceType<T>> {
  return new Module()
}

export * from './axios'
export * from './compiler'
