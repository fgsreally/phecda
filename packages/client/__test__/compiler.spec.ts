import { describe, expect, it } from 'vitest'
import type { P } from 'phecda-server'
import { Compiler } from '../src'
describe('fakeController', () => {
  it('compiler will create fake class', () => {
    const meta: P.Meta[] = [
      {
        name: 'Controller1',
        method: 'add',
        tag: '1',
        route: {
          route: '/a',
          type: 'post' as const,
        },
        header: {},
        guards: [],
        interceptors: [],
        middlewares: [],
        params: [{
          type: 'query',
          key: 'a',
          index: 0,
        }, {
          type: 'body',
          key: 'b',
          index: 1,
        }, {
          type: 'params',
          key: 'c',
          index: 2,
        }, {
          type: 'params',
          key: 'd',
          index: 3,
        }],
      },
      {
        name: 'Controller2',
        method: 'add',
        tag: '2',
        header: {},
        guards: [],
        interceptors: [],
        middlewares: [],
        route: {
          route: '/a',
          type: 'post' as const,
        },
        params: [{
          type: 'query',
          key: 'a',
          index: 0,
        }, {
          type: 'body',
          key: 'b',
          index: 1,
        }],
      },
    ]
    const faker = new Compiler()

    for (const i of meta)
      faker.addMethod(i)

    expect(faker.getContent()).toMatchSnapshot()
  })

  it('when key doesn\' exist', () => {
    const meta: P.Meta[] = [
      {
        name: 'Controller1',
        method: 'add',
        tag: '1',
        route: {
          route: '/a',
          type: 'post' as const,
        },
        header: {},
        guards: [],
        interceptors: [],
        middlewares: [],
        params: [{
          type: 'query',
          key: 'a',
          index: 0,
        }, {
          type: 'body',
          key: '',
          index: 1,
        }, {
          type: 'params',
          key: 'c',
          index: 2,
        }, {
          type: 'params',
          key: 'd',
          index: 3,
        }],
      },

    ]
    const faker = new Compiler()

    for (const i of meta)
      faker.addMethod(i)

    expect(faker.getContent()).toMatchSnapshot()
  })
})
