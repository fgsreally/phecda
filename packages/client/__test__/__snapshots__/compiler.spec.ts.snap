// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fakeController > compiler will create fake class 1`] = `
"
        export class Controller1{
            
    add(arg0,arg1,arg2,arg3){
const ret={tag:\\"1-add\\",body:{},headers:{},query:{},params:{},method:\\"post\\",url:\\"/a\\"}
if(arg0!==undefined&&arg0!==null){ret.query['a']=arg0
}
if(arg1!==undefined&&arg1!==null){ret.body['b']=arg1
}
if(arg2!==undefined&&arg2!==null){ret.params['c']=arg2
ret.url=ret.url.replace('{{c}}',arg2)}
if(arg3!==undefined&&arg3!==null){ret.params['d']=arg3
ret.url=ret.url.replace('{{d}}',arg3)}

return ret
    }
    
            }
        export class Controller2{
            
    add(arg0,arg1){
const ret={tag:\\"2-add\\",body:{},headers:{},query:{},params:{},method:\\"post\\",url:\\"/a\\"}
if(arg0!==undefined&&arg0!==null){ret.query['a']=arg0
}
if(arg1!==undefined&&arg1!==null){ret.body['b']=arg1
}

return ret
    }
    
            }"
`;

exports[`fakeController > when key doesn' exist 1`] = `
"
        export class Controller1{
            
    add(arg0,arg1,arg2,arg3){
const ret={tag:\\"1-add\\",body:{},headers:{},query:{},params:{},method:\\"post\\",url:\\"/a\\"}
if(arg0!==undefined&&arg0!==null){ret.query['a']=arg0
}
if(arg1!==undefined&&arg1!==null){ret.body=arg1
}
if(arg2!==undefined&&arg2!==null){ret.params['c']=arg2
ret.url=ret.url.replace('{{c}}',arg2)}
if(arg3!==undefined&&arg3!==null){ret.params['d']=arg3
ret.url=ret.url.replace('{{d}}',arg3)}

return ret
    }
    
            }"
`;
